client/app/bundles/course/assessment/question/programming/components/OnlineEditorCppView.jsx

```diff
@@ -0,0 +1,515 @@
+    deleteCppTestCase: PropTypes.func.isRequired,
+    updateCppNewDataFile: PropTypes.func.isRequired,
+    deleteCppNewDataFile: PropTypes.func.isRequired,
+    deleteCppExistingDataFile: PropTypes.func.isRequired,
+  }),
+  isLoading: PropTypes.bool.isRequired,
+  autograded: PropTypes.bool.isRequired,
+  intl: intlShape.isRequired,
+};
+
+const contextTypes = {
+  muiTheme: React.PropTypes.object.isRequired,
+};
+
+export function validation(data, pathOfKeysToData, intl) {
+  const errors = [];
+  const cppData = data.getIn(pathOfKeysToData);
+  if (data.getIn(['question', 'autograded'])) {
+    let testsCount = 0;
+
+    ['public', 'private', 'evaluation'].forEach((type) => {
+      const testCases = cppData.getIn(['test_cases', type]);
+      testsCount += testCases.size;
+
+      testCases.forEach((testCase, index) => {
+        const testCaseError = {};
+        let hasError = false;
+
+        if (testCase.get('expression').trim() === '') {
+          testCaseError.expression =
+            intl.formatMessage(translations.cannotBeBlankValidationError);
+          hasError = true;
+        }
+        if (testCase.get('expected').trim() === '') {
+          testCaseError.expected =
+            intl.formatMessage(translations.cannotBeBlankValidationError);
+          hasError = true;
+        }
+
+        if (hasError) {
+          errors.push({
+            path: pathOfKeysToData.concat(['test_cases', type, index, 'error']),
+            error: testCaseError,
+          });
+        }
+      });
+    });
+
+    if (testsCount === 0) {
+      errors.push({
+        path: pathOfKeysToData.concat(['test_cases', 'error']),
+        error: intl.formatMessage(translations.noTestCaseErrorAlert),
+      });
+    }
+  }
+
+  return errors;
+}
+
+class OnlineEditorCppView extends React.Component {
+
+  static getInputName(field) {
+    return `question_programming[${field}]`;
+  }
+
+  static getTestInputName(type, field) {
+    return `question_programming[test_cases][${type}][][${field}]`;
+  }
+
+  codeChangeHandler(field) {
+    return e => this.props.actions.updateCppCodeBlock(field, e);
+  }
+
+  newDataFileChangeHandler(index) {
+    return (e) => {
+      const files = e.target.files;
+      const filename = files.length === 0 ? null : files[0].name;
+      this.props.actions.updateCppNewDataFile(filename, index);
+    };
+  }
+
+  testCaseDeleteHandler(type, index) {
+    return (e) => {
+      e.preventDefault();
+
+      if (!this.props.isLoading) {
+        this.props.actions.deleteCppTestCase(type, index);
+      }
+    };
+  }
+
+  testCaseCreateHandler(type) {
+    return (e) => {
+      e.preventDefault();
+
+      if (!this.props.isLoading) {
+        this.props.actions.createCppTestCase(type);
+      }
+    };
+  }
+
+  renderExistingDataFiles = () => {
+    if (this.props.data.get('data_files').size === 0) {
+      return null;
+    }
+
+    const formatBytes = (bytes, decimals) => {
```

Comment:
> Seems to only be specific to the rendering of data files though, think it would be better to keep it within the function itself.

Actually I can use it this for the file upload component to specify the size of the uploaded file. But no worries, this part can be done later.